KAFKA-5142: Add Connect support for message headers (KIP-145)

rhauch opened the pull request:
**[KIP-145](https://cwiki.apache.org/confluence/display/KAFKA/KIP-145+-+Expose+Record+Headers+in+Kafka+Connect) has been accepted, and this PR implements KIP-145 except without the SMTs.**

Changed the Connect API and runtime to support message headers as described in [KIP-145](https://cwiki.apache.org/confluence/display/KAFKA/KIP-145+-+Expose+Record+Headers+in+Kafka+Connect).

The new `Header` interface defines an immutable representation of a Kafka header (key-value pair) with support for the Connect value types and schemas. This interface provides methods for easily converting between many of the built-in primitive, structured, and logical data types.

The new `Headers` interface defines an ordered collection of headers and is used to track all headers associated with a `ConnectRecord` (and thus `SourceRecord` and `SinkRecord`). This does allow multiple headers with the same key. The `Headers` contains methods for adding, removing, finding, and modifying headers. Convenience methods allow connectors and transforms to easily use and modify the headers for a record.

A new `HeaderConverter` interface is also defined to enable the Connect runtime framework to be able to serialize and deserialize headers between the in-memory representation and Kafka’s byte[] representation. A new `SimpleHeaderConverter` implementation has been added, and this serializes to strings and deserializes by inferring the schemas (`Struct` header values are serialized without the schemas, so they can only be deserialized as `Map` instances without a schema.) The `StringConverter`, `JsonConverter`, and `ByteArrayConverter` have all been extended to also be `HeaderConverter` implementations. Each connector can be configured with a different header converter, although by default the `SimpleHeaderConverter` is used to serialize header values as strings without schemas.

Unit and integration tests are added for `ConnectHeader` and `ConnectHeaders`, the two implementation classes for headers. Additional test methods are added for the methods added to the `Converter` implementations. Finally, the `ConnectRecord` object is already used heavily, so only limited tests need to be added while quite a few of the existing tests already cover the changes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation matches KIP-145
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-5142: Add message headers to Connect API (KIP-145)

Changed the Connect API to add message headers as described in KIP-145.

The new `Header` interface defines an immutable representation of a Kafka header (name-value pair) with support for the Connect value types and schemas. Kafka headers have a string name and a binary value, which doesn’t align well with Connect’s existing data and schema mechanisms. Thus, Connect’s `Header` interface provides methods for easily converting between many of the built-in primitive, structured, and logical data types. And, as discussed below, a new `HeaderConverter` interface is added to define how the Kafka header binary values are converted to Connect data objects.

The new `Headers` interface defines an ordered collection of headers and is used to track all headers associated with a `ConnectRecord`. Like the Kafka headers API, the Connect `Headers` interface allows storing multiple headers with the same key in an ordered list. The Connect `Headers` interface is mutable and has a number of methods that make it easy for connectors and transformations to add, modify, and remove headers from the record, and the interface is designed to allow chaining multiple mutating methods.

The existing constructors and methods in `ConnectRecord`, `SinkRecord`, and `SourceRecord` are unchanged to maintain backward compatibility, and in these situations the records will contain an empty `Headers` object that connectors and transforms can modify. There is also an additional constructor that allows an existing `Headers` to be passed in. A new overloaded form of `newRecord` method was created to allow connectors and transforms to create a new record with an entirely new `Headers` object.

A new `HeaderConverter` interface is also defined to enable the Connect runtime framework to be able to serialize and deserialize headers between the in-memory representation and Kafka’s byte[] representation.

Unit and integration tests are added for `ConnectHeader` and `ConnectHeaders`, the two implementation classes for headers. The `ConnectRecord` object is already used heavily, so only limited tests need to be added while quite a few of the existing tests already cover the changes. However, new unit tests were added for `SinkRecord` and `SourceRecord to verify the header behavior, including when the `newRecord` methods are called.

rhauch committed:
KAFKA-5142: Add message header converters to Connect API (KIP-145)

This is the second commit for the public Connect API changes for KIP-145, and deals primarily with `HeaderConverter` implementations.

Connect has three `Converter` implementations, `StringConverter`, `JsonConverter` and `ByteArrayConverter`. These were modified to also implement `HeaderConverter`, without changing any of the existing functionality.

Like many of our pluggable components in Connect, the `HeaderConverter` interface extends `Configurable` that allows implementations to expose a `ConfigDef` that describes the supported configuration properties, and a `config` method that can be used to initialize the component with provided configuration properties. The `StringConverter`, `JsonConverter` and `ByteArrayConverter` were changed to support these methods in a backward compatible manner. There are now `StringConverterConfig` and `JsonConverterConfig` classes that define the `ConfigDef` for the implementations; the `ByteArrayConverter` has no configuration properties and doesn't need a config class.

Note that the existing `Converter` interface has a special `config` signature with a parameter that sas whether the converter is being used for keys or values. This is different than the `Configurable.config` signature, so this commit adds new `ConverterConfig` abstract class that defines a `converter.type` property that can be used to set whether the converter is being used for keys, values, or headers. The existing `Converter` methods internally set this property based upon the supplied boolean parameter, so the default for `converter.type` can be `header`.

rhauch committed:
KAFKA-5142: Add message headers to Connect runtime (KIP-145)

This is the third commit for KIP-145 and changes the Connect runtime to support headers. Each Connect worker now configures a `HeaderConverter` for each connector task, in the same way it creates key and value `Converter` instances. This is entirely backward compatible, so that existing worker and connector configurations will work without changes. By default, the worker will use the `SimpleHeaderConverter` to serialize header values as strings and to deserialize them by inferring the schemas.

tedyu commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeader.java:
```diff
@@ -0,0 +1,403 @@
+
+    @Override
+    public Header rename(String key) {
+        Objects.requireNonNull(key, "Null header keys are not permitted");
```
Should check ! key.equals(this.key)

rhauch replied:
Yeah, I think that's a valid optimization.

tedyu commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,523 @@
+    private static final SchemaAndValue NULL_SCHEMA_AND_VALUE = new SchemaAndValue(null, null);
+
+    // This field is set lazily, but once set to a list it is never set back to null
+    private LinkedList<Header> headers;
```
this should be declared volatile (considering the double-checked locking below)

rhauch replied:
Agreed.

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/header/Headers.java:
```diff
@@ -0,0 +1,308 @@
+/**
+ * A mutable ordered collection of {@link Header} objects. Note that multiple headers may have the same {@link Header#key() key}.
+ */
+public interface Headers extends Iterable<Header> {
```
should we call class this something else to avoid confusion with `org.apache.kafka.common.header.Headers`?

rhauch replied:
No, I'd prefer to keep them the same name. There is no reason why a connector would ever use both Connect and Kafka header-related classes, and there are only a few places where Connect code uses both and it's easy enough to package-scope the types.

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/storage/SimpleHeaderConverter.java:
```diff
@@ -0,0 +1,728 @@
+        return value.replaceAll("\\\\", "\\\\\\\\").replaceAll("\"", "\\\\\"");
+    }
+
+    protected SchemaAndValue parse(Tokenizer tokenizer, boolean embedded) throws NoSuchElementException {
```
I am a bit worried that this parser could change the underlying data in small ways, and cause issues which become hard to debug. For example, the string "  0.11" would drop the initial spaces, and numbers in scientific notation might be translated into decimal (for example, 1E-2 becomes 0.01). I think this class is very useful, but do you think we want to have `org.apache.kafka.connect.converters.ByteArrayConverter` as the default converter?

rhauch replied:
Okay, what if we move all of this conversion logic into the `Header.valueAsX` methods? If we do that, I think we can use the `StringConverter` as the default and can avoid the `SimpleHeaderConverter` class altogether.

wicknicks replied:
Yes, that would make sense. The transformations would be explicit. Should we use `ByteArrayConverter` as the default?

rhauch replied:
After thinking about this more, the existing `StringConverter` uses `toString()` to serialize all objects, which means for arrays and maps the string representation contains no quotes around strings and so would be relatively difficult to parse. So I'm not convinced we can do without `SimpleHeaderConverter`.

Also, I'm not sure why it'd be an issue if the default header converter parser the values into nominal values, even if that results in slight variations. First, existing connectors don't handle headers now. Second, any connector that did require the exact representation can specify a non-default header converter. Third, it makes more sense to me to make the headers meaningful and useful by default.

wicknicks replied:
I think there is a lot of value in providing meaning to the headers. If we could make the transition explicit, that would be the best. Since connect sits between many different systems with non compatible data models, a translation may cause unexpected issues in production. For example, a long in Java is up to 2^64-1, but javascript would use double precision to handle big integers.

Regarding the problem of serializing maps, I agree with you (`StringConverter` and `ByteArrayConverter` won't work). 

Would it fine to think that if people are converting headers in their tasks or SMTs, then we can expect them to set `SimpleHeaderConverter`  as their converter class? And if they never care about what is done to the headers, then it should be correct to have set `ByteArrayConverter` as the default value for `header.converter` property in `WorkerConfig.java`?

rhauch replied:
Yes, it is certainly possible to use the `ByteArrayConverter` in either the worker config or the connector config, and when this is used the byte array in each header value will simply be passed directly into the `Headers` object for each record.

I'm still not convinced that this should be the *default* header converter, though.

rhauch committed:
KAFKA-5142: Changed Connect records to accept Iterable<Header> rather than only Headers (KIP-145)

rhauch committed:
KAFKA-5142: Corrected checkstyle and refactored SimpleHeaderConverter

rhauch committed:
KAFKA-5142: Corrected how ValueConversion serializes Struct

rhauch committed:
KAFKA-5142: Revised the design per the KIP-145 discussion

Changed the design of the public API to reflect the latest KIP-145 discussion. Also corrected several checkstyle and findbugs warnings.

rhauch committed:
KAFKA-5142: Additional changes to correct style and bugs

Removed the `Header.valueAsX()` methods and replaced them with `Values.convertToX(Schema, Object)` methods that can be used outside of headers. These methods handle converting between all of the primitives and logical types, as well as between string and arrays/maps.

rhauch committed:
KAFKA-5142: Mentioned Connect headers in the documentation and upgrade guide

rhauch commented:
Test results are unrelated:

```
java.io.IOException: No space left on device
	at java.base/java.io.RandomAccessFile.setLength(Native Method)
```

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeader.java:
```diff
@@ -0,0 +1,103 @@
+        if (obj == this) {
+            return true;
+        }
+        if (obj instanceof ConnectHeader) {
```
nit : we probably would get the same result if we just had the equals check that does key schema & value . Instead of 2 if conditions.

rhauch replied:
Agreed.

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeader.java:
```diff
@@ -0,0 +1,103 @@
+
+    private static final SchemaAndValue NULL_SCHEMA_AND_VALUE = new SchemaAndValue(null, null);
+
+    private static final String TRUE_WORD_LITERAL = "true";
```
I don't see these literals used anywhere

rhauch replied:
Nope, they're not. Removing them.

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+
+    @Override
+    public Headers add(Header header) {
+        if (header == null) {
```
may be use  Objects.requireNonNull

rhauch replied:
ack

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+     * @param schemaAndValue the schema and value pair
+     * @throws DataException if the schema is not compatible with the value
+     */
+    void checkSchemaMatches(SchemaAndValue schemaAndValue) {
```
would SchemaAndValue be a better place to have this method?

rhauch replied:
For some time it was not obvious to me that `SchemaAndValue` is not part of the public API, so my concern about that would be that it'd have to become a public method on `SchemaAndValue`, which is currently just a simple container. Here, we can keep it out of the API altogether.

mageshn replied:
Got it.  May be we could have something like SchemaAndValueUtils to include such utils for it. Just  didn't feel great seeing this method in the ConnectHeaders class.

ewencp replied:
`SchemaAndValue` is definitely public, it's used in some public APIs. Agree that there might be some handy utilities that could be colocated. I'd wait to put them there (and deal with any potential KIP fallout) until we know there are multiple use cases. atm, we don't have multiple use cases.

However, @rhauch, one question would be whether `ConnectSchema.validateValue` could work here instead of these methods?

rhauch committed:
KAFKA-5142: Addressed suggestions from reviews.

Removed unused constants and unnecessary block, and added optimization for `Header.rename(newKey)` when the new key is the same as the old key.

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+    }
+
+    @Override
+    public Headers retainLatest(String key) {
```
retainLatest() and this method have a lot in common. We could potentially refactor it, but not too concerned if its left as-is.

rhauch replied:
The duplication is not ideal, but since the logic varies deep within the loop it's actually relatively complex to implement with a single method using Java 7, and we'd actually end up with something that's more complex. With Java 8 it'd be super easy, so I guess I'd prefer to leave it as is.

ewencp replied:
i think leaving as is should be fine atm, and tbh at least they are both close enough together to be easily modified together. if we think this is useful enough, i'd file a jira dependent on the jdk8 update so we can follow up.

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+    }
+
+    @Override
+    public Headers apply(HeaderTransform transform) {
```
similar comment as retainLatest

rhauch replied:
same response as above. :-)

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+        if (obj == this) {
+            return true;
+        }
+        if (obj instanceof ConnectHeaders) {
```
probably not required to do a special logic for ConnectHeaders. The equals check using iterator below should probably be suffice.

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+     * @throws IllegalArgumentException if the supplied key is null
+     */
+    private void checkKey(String key) {
+        if (key == null) {
```
Maybe use  Objects.requireNonNull

mageshn commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,534 @@
+     */
+    private void checkSchemaType(Schema schema, Type type) {
+        if (schema.type() != type) {
+            throw new DataException("The schema for a list must be of type " + type);
```
Exception message doesn't look right (the word "list").

mageshn reviewed (commented):
Looks Great. Thanks Randall.

rhauch committed:
KAFKA-5142: Additional fixes per reviews.

rhauch commented:
@mageshn, thanks for the review. I added a commit to fix several of the issues you identified. I left the `retainLatest` and `apply` methods as is, since refactoring would be non-trivial and not as straightforward. (See comment above for details.)

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+import java.util.TimeZone;
+
+/**
+ * Utility for converting values into a different form.
```
Sounds like a tautology to the name of the class. A bit more detail would be nice. Even _Connect record values_ could be more specific.

rhauch replied:
I'll add more detail.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+
+
+    // This field is set lazily, but once set to a list it is never set back to null
+    private volatile LinkedList<Header> headers;
```
Why not use `List` (interface)?

rhauch replied:
simply because we need to iterate backward and forward.

ewencp commented on checkstyle/suppressions.xml:
```diff
@@ -116,6 +120,12 @@
               files="JsonConverter.java"/>
     <suppress checks="NPathComplexity"
               files="DistributedHerder.java"/>
+    <suppress checks="NPathComplexity"
```
why are we hitting these for cases like ConnectHeader? doesn't look that complex. this is a big patch, with a bunch of new code so I was expecting some of these, but it seems like we're adding quite a lot of new exceptions.

rhauch replied:
`ConnectHeader` needed a suppression when it had all of the conversion methods, but since those are gone I can remove that suppression. `ConnectHeaders` and `Values` still are complex enough that they need a suppression.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/connector/ConnectRecord.java:
```diff
@@ -34,18 +39,33 @@
         this.valueSchema = valueSchema;
         this.value = value;
         this.timestamp = timestamp;
+        if (headers == null) {
```
not really critical, but this seems like a departure since we don't validate, e.g., `topic`. Why would you even use the version with `headers` as an argument if you weren't going to pass it in? another option would just be adding the non-null validation.

rhauch replied:
I guess I was just trying to be safe in case someone is passing in a reference they didn't know is null. Seemed easy enough to handle, especially since not handling such a case means the connector task dies. But I can clean it up a bit since the `ConnectHeaders` constructor can take a null value.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+    private static final String TRUE_LITERAL = Boolean.TRUE.toString();
+    private static final String FALSE_LITERAL = Boolean.TRUE.toString();
+    private static final Charset UTF_8 = StandardCharsets.UTF_8;
+    private static final long MILLIS_PER_DAY = 24 * 60 * 60 * 1000;
```
I'm fine with this but some people prefer to do this via TimeUnits

rhauch replied:
`Time` and `Date` in the same package do exactly this. Is it better to be consistent but use the constants, or to use TimeUnits here? I went with consistent but am happy to change it.

ewencp replied:
either way is fine, i don't really java which is probably way `Time` and `Date` do this. was just noting since it is `TimeUnits` plenty of other places

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+    private static final String ARRAY_END_DELIMITER = "]";
+    private static final String MAP_BEGIN_DELIMITER = "{";
+    private static final String MAP_END_DELIMITER = "}";
+    private static final int IS_8601_DATE_LENGTH = ISO_8601_DATE_FORMAT_PATTERN.length();
```
I think these are accidentally `IS_` instead of `ISO_`.

rhauch replied:
fixed.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+    }
+
+    /**
+     * Convert the specified value to an {@link Type#ARRAY} value.
```
It might be worth documenting what happens recursively and if the schema is obeyed. e.g., if I pass in a schema for `Array<Double>`, is that what I get back? If I pass in `null` for the schema, what is the type of the elements?

rhauch replied:
Updated the doc of this and other methods to hopefully be more clear. Basically, with `convertToList`, `convertToMap`, and `convertToStruct`, the schema is not used, but it is used in many of the numeric conversions when the supplied values are logical values. I kept it in the signature (a) to be consistent with the other forms, and (b) in case we need it for more complex conversions in the future.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+     * @param value the value whose schema is to be inferred; may be null
+     * @return the inferred schema, or null if the value is null or no schema could be inferred
+     */
+    public static Schema inferSchema(Object value) {
```
do we have this stuff elsewhere already? anywhere other than the converters? just curious if we can reuse some code instead of duplicating since I'd imagine this type inference is pretty common.

ewencp replied:
actually, now I remember, it looks familiar because I have basically the same code in my schema inference SMT that I just have sitting on some branch but would have needed an SMT. I guess if we end up with that SMT we can probably just leverage this code.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+    public Headers add(Header header) {
+        Objects.requireNonNull(header, "Unable to add a null header.");
+        if (this.headers == null) {
+            synchronized (this) {
```
I'm a little bit confused regarding what are the concurrency (thread-safety) assumptions that we have for this class. 

1) If it's meant to be used only by a single thread at a time, I'd expect the declaration of `headers` as volatile to suffice for the lazy initialization, and the object would be safe for hand-off between threads (but not concurrent access). 
2) If the pattern of double-checked locking (which I don't love even in its correct post-JDK5 form, but that's not the issue here) is necessary, because there might be concurrent access of a `ConnectHeaders` object, I'd assume this class is broken, because iterators are shared and I'd expect `ConcurrentModificationException`s to occur (e.g. a header item is added by a thread while the headers list is traversed by another thread). 

I'm pretty sure our case corresponds to 1). But I need to ask. If that's correct, I don't think we need the `synchronized` block around `headers` initialization and double-checked locking. By taking a look at the implementation of `RecordHeaders` which seems to share similarities with this class, there's no `synchronized` access to its `volatile` field (the list there is `final` but there's a boolean that is `volatile`).

I hope I'm not missing something.

rhauch replied:
The whole point of this is to not actually create the list if there are no headers on a record. I'd expect that to be by far the prevalent case, so I thought it worthwhile to deal with that.

However, as you suggest, we should only have single threaded access, so the use of `volatile` and double-checked locking is, to say the least, ridiculous.

kkonstantine replied:
I get the intention for lazy initialization. Asked generally since I wanted to be 100% sure about the expectations w.r.t. concurrency. 

I think it's best to keep `volatile` since the field is not final and might be accessed in sequence by different threads, but `synchronized` can be skipped if there's no concurrent execution by different threads.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+
+    @Override
+    public boolean isEmpty() {
+        return this.headers == null ? true : headers.isEmpty();
```
I find a little bit confusing that in some methods `this.headers` is uses although there's no conflict with arguments (as it happens here), while in other methods `headers` is used without `this.` to refer to the member field (e.g. method `lastWithName`). Should we avoid `this` outside the constructor altogether?

ewencp replied:
the norm (ideally) for AK is to only use `this.` as needed to handle masking. i'm sure there are exceptions, but almost universally the use case is for constructors to avoid awkward parameter naming in order to avoid conflicting with a convenient member variable name.

rhauch replied:
Well, at least we should be consistent within a compilation unit, so I'll fix that.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+    }
+
+    /**
+     * Check that the key is not null
```
Feels a bit strange that javadoc is added for trivial `private` methods as this one, but not for `protected` methods elsewhere. Maybe we want to be more specific with javadoc requirements.

rhauch replied:
I tend to put JavaDoc when the method throws an exception and it's not completely obvious when trying to use that method when that exception may occur. Hopefully that saves a step in the IDE when trying to use a method. And of course that cascades to needing full JavaDoc.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+     * @param schemaAndValue the schema and value pair
+     * @throws DataException if the schema is not compatible with the value
+     */
+    void checkSchemaMatches(SchemaAndValue schemaAndValue) {
```
I've mentioned before, but I find package-private scope unintuitive. If this is like that only for testing, maybe a comment would help, otherwise I'd suggest `protected` to be used.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+        }
+
+        protected Header makeNext() {
+            while (true) {
```
Infinite loop doesn't seem to make sense.
Should this be re-written as?
```java
protected Header makeNext() {
  while (original.hasNext()) { 
    Header header = original.next();
    if (!header.key().equals(key)) {
      continue;     
    }           

    return header;
  }         
  return this.allDone();
}
```

rhauch replied:
Yeah, I stole this from another part of the code and didn't bother to refactor.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/storage/SimpleHeaderConverter.java:
```diff
@@ -0,0 +1,85 @@
+ */
+public class SimpleHeaderConverter implements HeaderConverter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SimpleHeaderConverter.class);
```
let's keep it `log` everywhere.

kkonstantine commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+ */
+public class Values {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Values.class);
```
Let's keep it `log` everywhere. Since we have the checkstyle suppression anyways (still I'd prefer `LOG` :) )

rhauch replied:
sure.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+            return Schema.FLOAT64_SCHEMA;
+        }
+        if (value instanceof byte[]) {
+            return Schema.BYTES_SCHEMA;
```
we should check for `ByteBuffer` as well, that's supported for the `byte[]` type

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -109,7 +115,7 @@ public void ensureValid(String name, Object value) {
                         }
                     }
-                }), Importance.LOW, TRANSFORMS_DOC, TRANSFORMS_GROUP, 6, Width.LONG, TRANSFORMS_DISPLAY);
+                }), Importance.LOW, TRANSFORMS_DOC, TRANSFORMS_GROUP, 7, Width.LONG, TRANSFORMS_DISPLAY);
```
Do we need a counter?

rhauch replied:
I was trying to not touch even more lines of code. But, yes, that's probably best.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -207,6 +208,25 @@ public Converter newConverter(String converterClassOrAlias, AbstractConfig confi
+            throw new ConnectException(
+                    "Failed to find any class that implements HeaderConverter and which name matches "
+                            + converterClassOrAlias
+                            + ", available connectors are: "
```
let's fix it here and on the Exception message above. "available converters are: "

rhauch replied:
ok

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+            if (toSchema.isOptional()) {
+                return null;
+            }
+            throw new DataException("Unable to convert a null value to a non-null value");
```
this message should probably end with "to a non-optional schema" as that's clearer what the real source of the problem is. (or "required" instead of "non-optional" though I can't remember how commonly we use that term in Connect.)

kkonstantine commented on docs/upgrade.html:
```diff
@@ -74,6 +74,7 @@ <h5><a id="upgrade_110_notable" href="#upgrade_110_notable">Notable changes in 1
 	Kafka Streams retries and can <a href="/{{version}}/documentation/streams/developer-guide/config-streams">configure<a/>
 	fine-grained timeouts (instead of hard coded retries as in older version).</li>
     <li>Kafka Streams rebalance time was reduced further making Kafka Streams more responsive.</li>
+    <li>Kafka Connect now supports message headers in both sink and source connectors, and to manipulate them via simple message transforms. Connectors must be changed to explicitly use them. A new "HeaderConverter" is introduced to control how headers are (de)serialized, and the new "SimpleHeaderConverter" is used by default to use string representations of values.</li>
```
Should we use `<code>` vs double-quotes?

rhauch replied:
fixed.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+            throw new DataException("Unable to convert a null value to a non-null value");
+        }
+        switch (toSchema.type()) {
+            case BYTES:
```
i won't block merging this on this, but all of this code is where we could probably drop npath and cyclomatic complexities and get rid of those checkstyle issues. (`JsonConverter` has the same issues from before we had those in place, so no shame here, mea culpa too :) )

the `instanceof` checks are difficult to fix because you can't trivially turn it into a map lookup and dispatch to that function, but the top level cases can be broken into their own methods, each with lower complexity. (admittedly, even then we may hit our limits due to cases where we need to handle a lot of different types in unique ways.)

rhauch replied:
Right, but that while that does end up making each method less complex, the expansion into many methods makes the class harder to really understand the potential paths. Seems like we'd just be playing games with checkstyle and not _really_ simplifying the code.

ewencp replied:
yeah, people differ in their opinions about granularity and understandability. I think I personally align more w/ you, where just inlining stuff seems more understandable overall.

if others agree, maybe things like cyclomatic complexity can simply have increased limits instead of requiring breaking these things up more. that said, this code is definitely an extreme example -- i had to work a bit to track alignment to make sure I understood where, and at what depth, various blocks were.

kkonstantine reviewed (commented):
Nice milestone. I left a few comments. Thanks @rhauch !

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+        }
+        switch (toSchema.type()) {
+            case BYTES:
+                if (Decimal.LOGICAL_NAME.equals(toSchema.name())) {
```
not sure its anything to be fixed here, but I had been hoping to keep logical types as obvious pre- or post- processing steps (and where possible, dispatch to handlers via simple map lookups). I would need to think more about how to make that work in this case, but something to consider, even if just for a future refactoring.

rhauch replied:
Agreed, though that would require adding conversion methods to all logical types and then registering them somewhere. I'd prefer to refactor in the future, given the already monstrous size of this PR.

ewencp replied:
oh yes, refactoring in the future preferred. i only mentioned in case the original author had an easy fix :)

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                        return new BigDecimal(value.toString()).doubleValue();
+                    }
+                }
+                if (value instanceof byte[]) {
```
same comment about ByteBuffer here

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                }
+                if (value instanceof String) {
+                    SchemaAndValue parsed = parseString(value.toString());
+                    if (parsed.value() instanceof Boolean) {
```
Still working through how all of this works, but this seems awkward -- why would `parseString` return a `SchemaAndValue`, but then we'd be doing regular java type checking on the `value()`?

rhauch replied:
Line 405 could instead be:

    if (parsed.schema().type() == Type.BOOLEAN) {

but that didn't really seem that much better. 

The string parser always returns a SchemaAndValue, largely because the schema of any nested values needs to be known for the schema of the container (e.g., an array of boolean values vs an array of int8 values). We don't need that here because these are unwrapped primitives.

ewencp replied:
fair enough, just felt a bit awkward. with very large chunks of code like this, it is hard to understand the reasoning without having tried to write it / deeply modify it yourself.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                        return parsed.value();
+                    }
+                }
+                return asLong(value, fromSchema, null) == 1L ? Boolean.TRUE : Boolean.FALSE;
```
i think this is an age old question, but do we want to be restrictive on what values result in `true` or select exactly one value for `false`. perhaps i'm just stuck too much in C-mode, but true for anything non-zero is what's intuitive for me.

rhauch replied:
ack.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                if (value instanceof Byte) {
+                    return value;
+                }
+                return (byte) asLong(value, fromSchema, null);
```
do we actually want to use these conversions that just wrap/truncate values or should we be throwing errors if they are out of range? the latter seems better to me, but probably complicates the code as we can't rely only on `asLong` and casting...

rhauch replied:
I could go either way. Maybe we should do that as a followup PR?

ewencp replied:
we can follow up, but it is a significant behavior so we should lock it down before releasing this. otherwise the behavior change is a kip

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                return number.longValue();
+            }
+            if (value instanceof String) {
+                return new BigDecimal(value.toString()).longValue();
```
`value` is already a `String`, why would we need `value.toString()` here?

rhauch replied:
simply saves a cast.

ewencp replied:
ack, honestly not sure if saving the cast or saving the vtable lookup is better

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                return number.doubleValue();
+            }
+            if (value instanceof String) {
+                return new BigDecimal(value.toString()).doubleValue();
```
same question about unnecessary `toString` here

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+                sb.append(value);
+            }
+        } else if (value instanceof byte[]) {
+            value = new String((byte[]) value, UTF_8);
```
Is this right? I thought byte[] would be encoded as base64? this call is guaranteed to be successful, but it will just replace "malformed" data with the default replacement string, which almost certainly is not what a user would want.

rhauch replied:
Yeah, that's what we discussed on the KIP discussion; I just neglected to make that change. Nice catch.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1168 @@
+        }
+    }
+
+    protected static Object project(Schema desiredSchema, Object value) {
```
This seems a bit repetitive, but more restrictive, as what `convertTo` is doing. Do we actually need both versions? Could we just unify the code with some flag for `project` vs `promote`?

rhauch replied:
No, we don't need the two different forms. That was leftover from an earlier state. I've added another commit that cleans this up.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+ */
+public class ConnectHeaders implements Headers {
+
+    private static final int EMPTY_HASH = Objects.hash(new LinkedList<>());
```
what are `EMPTY_HASH` and `EMPTY_ITERATOR` trying to optimize? one case mentions immutability, but in general why would these be shared? i could see some concerns with a shared, standard set of headers being used across multiple records, then being passed to transforms, but wouldn't that always require copying the data, not just in the empty cases?

rhauch replied:
I expect that most records will not have headers, and so `ConnectHeaders` currently allows the internal list to be null. Using `EMPTY_HASH` was a simple way to ensure that the hash code is the same for both situations. 

And, if it is true that most records won't have headers, then why construct a new Iterator every time it's needed? All of those cases (and only those cases) can share a single immutable `EMPTY_ITERATOR`. Since an iterator allows removing items (which we don't have in this case) but not adding them, there's no reason to have a per-usage `Iterator` instance.

ewencp replied:
might be true now, probably not true long term. also probably depends on where this is used - in a transformation for a source connector, it's likely for the foreseeable future that the headers are empty; for a sink connector, anywhere people have started using headers it is very unlikely they are empty.

the optimization is fine, i just watch for these things as they complicate the code and if they appear in the first version of code, usually aren't backed up by real data suggesting they are valuable.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+    private static final int EMPTY_HASH = Objects.hash(new LinkedList<>());
+
+    // Shared, immutable instance with a null schema and null value
+    private static final SchemaAndValue NULL_SCHEMA_AND_VALUE = new SchemaAndValue(null, null);
```
I've seen this a few places -- `SchemaAndValue` already has `SchemaAndValue.NULL` field which does the same thing -- no need to repeat a bunch of times in a bunch of classes.

rhauch replied:
Gotcha.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+    }
+
+    public ConnectHeaders(Iterable<Header> original) {
+        if (original != null) {
```
style nit: if the entire body is surrounded in a conditional, it's usually more readable to just check the negation and return, then reduce indentation with the rest of the body.

no real need to fix here, just a style thing to watch out for moving forward.

rhauch replied:
Oh, I actually prefer returning immediately, but thought that style was less preferred. Happy to change it.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+    // This field is set lazily, but once set to a list it is never set back to null
+    private volatile LinkedList<Header> headers;
+
+    public ConnectHeaders() {
```
are we trying to optimize something by not having `headers` always be non-null? seems like we complicate (and increase riskiness) of the implementation here by not just always assuming `headers` is non-null. Is the issue the mutability of this class and the `headers` collection?

rhauch committed:
KAFKA-5142: Fixes per code review.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+        if (value == null) {
+            return add(key, null, null);
+        }
+        // Check that this is a date ...
```
super-nit: this should say `Check that this is a Decimal`, not `date`. repeated below for time and timestamp, but slightly less egregiously :)

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+
+    @Override
+    public String toString() {
+        return "ConnectHeader(headers=" + (headers != null ? headers : "") + ")";
```
nit: should be `ConnectHeaders`. *probably* would be easy to figure out, but better to just get it right :)

rhauch replied:
good catch.

rhauch committed:
KAFKA-5142: More corrections per reviews.

rhauch commented:
Added two commits that address comments from @ewencp and @kkonstantine, respectively.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java:
```diff
@@ -0,0 +1,524 @@
+            if (value != null) {
+                switch (schema.type()) {
+                    case BYTES:
+                        if (value instanceof byte[])
```
Just a reminder, `ByteBuffer` again :)

rhauch committed:
KAFKA-5142: Removed method that was similar to but more constrained than convertTo

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/storage/ConverterConfig.java:
```diff
@@ -0,0 +1,59 @@
+public abstract class ConverterConfig extends AbstractConfig {
+
+    public static final String TYPE_CONFIG = "converter.type";
+    public static final String TYPE_DEFAULT = ConverterType.HEADER.getName();
```
this seems a weird default given that headers are new? wouldn't the default be empty/nothing since currently converters don't assume default types?

rhauch replied:
Ok, I'm fine with having null for the default. The reason I went this way was because of the non-standard `Converter.configure(Map<String, ?> configs, boolean isKey)` method that is currently used for all key and value converters, so right now the only thing that uses the `Configurable.configure(Map<String, ?> configs)` method is the header converter. However, that already sets the converter type, so I'm fine with not having a default.

rhauch replied:
Actually, on second thought, there's no reason to have a default. As mentioned above, the current code always sets it for key, value, and header converters, so having a default is actually incorrect.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -200,7 +201,26 @@ public Converter newConverter(String converterClassOrAlias, AbstractConfig confi
+            klass = pluginClass(
+                    delegatingLoader,
+                    converterClassOrAlias,
+                    HeaderConverter.class
```
are we continuing to add layers (and time) loading things? might be the right tradeoff now, but if we keep adding slowdown, we should consider spending time optimizing it as well.

rhauch replied:
Yes, we are, but right now there's no other way to find plugins for `HeaderConverter` implementations (that don't also implement `Converter`). There's already [KAFKA-6503](https://issues.apache.org/jira/browse/KAFKA-6503) that I'd like to address in 1.1.

ewencp replied:
sounds good. hadn't seen that bug, but is something i had also mentioned when we added the new plugin.path stuff. i would also very much like to see this.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java:
```diff
@@ -116,9 +124,6 @@ public void testStartAndStopConnector() throws Exception {
         expectStartStorage();
 
         // Create
-        Connector connector = PowerMock.createMock(Connector.class);
```
removing a bunch of boilerplate is great!

rhauch committed:
KAFKA-5142: More fixes based on code reviews.

rhauch commented:
Rebased due to several conflicts with changes already on `trunk`. Hopefully this will go green.

wicknicks commented on connect/api/src/test/java/org/apache/kafka/connect/source/SourceRecordTest.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test
+    public void shouldCreateSinkRecordWithEmtpyHeaders() {
```
I think you meant `shouldCreateSourceRecordWithEmtpyHeaders`?

ewencp commented:
LGTM. One good run, one failing on flaky core integration tests, and one that seems to be linking to the wrong job...

Merging to trunk for 1.1.0

wicknicks commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -739,6 +739,23 @@ public void testJsonSchemaCacheSizeFromConfigFile() throws URISyntaxException, I
+
+    @Test
+    public void stringHeaderToConnect() {
+        assertEquals(new SchemaAndValue(Schema.STRING_SCHEMA, "foo-bar-baz"), converter.toConnectHeader(TOPIC, "headerName", "{ \"schema\": { \"type\": \"string\" }, \"payload\": \"foo-bar-baz\" }".getBytes()));
```
nit: line is too long.

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -63,6 +64,15 @@
+                    " independent of connectors it allows any connector to work with any serialization format." +
+                    " Examples of common formats include JSON and Avro. By default, the SimpleHeaderConverter is used to serialize" +
+                    " header values to strings and deserialize them by inferring the schemas.";
+    public static final String HEADER_CONVERTER_CLASS_DEFAULT = SimpleHeaderConverter.class.getName();
```
since you have multiple implementations of HeaderConverter, should we add a list validator here listing them?

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1117 @@
+/*
```
this class has some fairly large methods. it might be worthwhile to break them into smaller methods for two reasons: (1) better readability; and (2) the JVM does throw in some better optimizations (inlining, for example) when there are small methods.

wicknicks commented on connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:
```diff
@@ -0,0 +1,1117 @@
+                }
+                // Missing either a comma or an end delimiter
+                if (COMMA_DELIMITER.equals(parser.previous())) {
+                    throw new DataException("Malformed array: missing element after ','");
```
Could you try running this test:
```
SchemaAndValue arr = Values.parseString("[1, 2, 3,,,]");
// expect an Exception.
```
I think it returns incorrect values. The trailing commas return the value of `parser.original()` from [line 836](https://github.com/apache/kafka/pull/4319/files?diff=unified#diff-fbab5614eb6f1b3a75939c8fb7df6dc2R866). It should return something more local here.

wicknicks reviewed (commented):
Great work, @rhauch! Some minor comments and a possible bug.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/201/

nick-laqua-dragon commented:
Is there any plan to implement the SMT’s (as specified by KIP-145) as well ?
